### PR TITLE
System mouse cursor: Web

### PR DESCRIPTION
**Pull request by dkwingsmt**
# Description

This PR adds system mouse cursor to the web engine.

This PR adds only a limited pool of system cursors. More will be added when the system is set up.

# Related Issues

* Depends on https://github.com/flutter/flutter/pull/54171 to properly work, but neither one is required to build.
* Part of https://github.com/flutter/flutter/issues/31952

# Known issues

The mouse position flickers when the cursor changes.

**Update**: This is not a Flutter issue, but a general issue of macOS

Entering: 
image
Leaving:
image

### CODE REVIEW

**Comment by dkwingsmt** [comment]
@yjbanov @ferhatb Can you take a look, also do you have an idea about the flickering?

**Comment by dkwingsmt** [comment]
It seems to me that this issue is not related to Flutter but a general issue (I noticed it on https://css-tricks.com/almanac/properties/c/cursor/).

File: lib/web_ui/lib/src/engine/mouse_cursor.dart [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  MouseCursor._() {
+  }
+
+  static String _mapShapeCodeToCssValue(int shapeCode) {
+    // Shape codes are hard-coded identifiers for system cursors.
+    // 
+    // The shape code values must be kept in sync with flutter's
+    // rendering/mouse_cursor.dart
+    switch (shapeCode) {
+      case /* none */ 0x334c4a:
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/mouse_cursor.dart:31]: Can you define kShapeCodeX/Y/Z in engine and then use those codes in rendering/mouse_cursor instead?

Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/mouse_cursor.dart:31]: Is it possible to import engine code from framework at all? Also these constants have to be replicated to other platforms as well, including Android, macOS, GLFW, etc.

**Comment by ferhatb** [comment]
I noticed that resize cursors are missing from codes which are used very frequently for things like resizing dividers etc. Are you planning to add more ? The web list of cursors is probably a good set to look at.

**Comment by dkwingsmt** [comment]
Yes, all cursors from CSS will be supported in an upcoming PR. This PR is a first step that aims to set up the infrastructure, and only include a limited set of cursors to limit the scope.

File: lib/web_ui/lib/src/engine/mouse_cursor.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/mouse_cursor.dart:31]: They would be part of dart:ui , just like StrokeJoin or other consts/enums. @gspencergoog : how do you typically handle consts across the 2 repos?

Comment (gspencergoog) [reply lib/web_ui/lib/src/engine/mouse_cursor.dart:31]: Yeah, actually, I think Ferhat is right: we should define the codes just as contiguous integers, and define an enum that matches in the Dart code.

`StrokeJoin` defines an enum, and then in the C++/Java code, it defines constants that are used as indices into the `StrokeJoin.values` member.

Then the only things that need to be kept in sync are the enum (which is in the engine repo in dart:ui), and the constants (which are also in the engine repo).

You can also look at `ImageByteFormat` or `TileMode`, or any of the other enums we use that way in the engine.

**Comment by dkwingsmt** [comment]
The described bug is actually from macOS, and not related to Web whatsoever.

**Comment by nturgut** [comment]
I think Linux Fuchsia unit tests were a temporary breakage. I checked the LUCI , builds the current PRs are passing. May be we can sync to the head and re-run.

**Comment by ayushin** [comment]
Still present in `master` on Mac OSX Safari Version 13.1 (15609.1.20.111.8) 

`Overflow on channel: flutter/mousecursor.  Messages on this channel are being discarded in FIFO fashion.  The engine may not be running or you need to adjust the buffer size if of the channel.
SE — channel_buffers.dart:144`

`[✓] Flutter (Channel master, 1.19.0-2.0.pre.143, on Mac OS X 10.15.4 19E287, locale en-CZ)`

**Pull request merged**